src/can.rs

```diff
@@ -0,0 +1,357 @@
+            filter: 0,
+            mask: 0x7FF,
+        }
+    }
+
+    /// Filter that accepts all extended CAN IDs.
+    pub fn extended_allow_all() -> Self {
+        Self::Extended {
+            filter: 0,
+            mask: 0x1FFFFFFF,
+        }
+    }
+}
+
+/// CAN abstraction
+pub struct CanBus<TX: OutputPin, RX: InputPin> {
+    tx: PhantomData<TX>,
+    rx: PhantomData<RX>,
+}
+
+impl<TX: OutputPin, RX: InputPin> CanBus<TX, RX> {
+    pub fn new(tx: TX, rx: RX, timing: Timing, filter: Filter) -> Result<Self, EspError> {
+        let general_config = twai_general_config_t {
+            mode: twai_mode_t_TWAI_MODE_NORMAL,
+            tx_io: tx.pin(),
+            rx_io: rx.pin(),
+            clkout_io: -1,
+            bus_off_io: -1,
+            tx_queue_len: 5,
+            rx_queue_len: 5,
+            alerts_enabled: TWAI_ALERT_NONE,
+            clkout_divider: 0,
+            intr_flags: ESP_INTR_FLAG_LEVEL1 as i32,
+        };
+
+        let timing_config = timing.timing();
+
+        // modify filter and mask to be compatible with TWAI acceptance filter
+        let (filter, mask) = match filter {
+            Filter::Standard { filter, mask } => ((filter as u32) << 21, !((mask as u32) << 21)),
+            Filter::Extended { filter, mask } => (filter << 3, !(mask << 3)),
+        };
+
+        let filter_config = twai_filter_config_t {
+            acceptance_code: filter,
+            acceptance_mask: mask,
+            single_filter: true,
+        };
+
+        esp!(unsafe { twai_driver_install(&general_config, &timing_config, &filter_config) })?;
+        esp!(unsafe { twai_start() })?;
+
+        Ok(Self {
+            tx: PhantomData,
+            rx: PhantomData,
+        })
+    }
+}
+
+impl<TX: OutputPin, RX: InputPin> embedded_hal::can::blocking::Can for CanBus<TX, RX> {
+    type Frame = Frame;
+    type Error = embedded_hal::can::ErrorKind;
+
+    fn transmit(&mut self, frame: &Self::Frame) -> Result<(), Self::Error> {
+        esp!(unsafe { twai_transmit(&frame.0, portMAX_DELAY) })
+            .map_err(|_| embedded_hal::can::ErrorKind::Other)
+    }
+
+    fn receive(&mut self) -> Result<Self::Frame, Self::Error> {
+        let mut rx_msg = twai_message_t {
+            ..Default::default()
+        };
+
+        match esp_result!(unsafe { twai_receive(&mut rx_msg, portMAX_DELAY) }, ()) {
+            Ok(_) => Ok(Frame(rx_msg)),
+            Err(_) => Err(embedded_hal::can::ErrorKind::Other),
+        }
+    }
+}
+
+impl<TX: OutputPin, RX: InputPin> embedded_hal::can::nb::Can for CanBus<TX, RX> {
+    type Frame = Frame;
+    type Error = embedded_hal::can::ErrorKind;
+
+    fn transmit(&mut self, frame: &Self::Frame) -> nb::Result<Option<Self::Frame>, Self::Error> {
+        // ESP_FAIL is i32 but others are u32
+        const _ESP_OK: i32 = ESP_OK as i32;
+        const _ESP_ERR_TIMEOUT: i32 = ESP_ERR_TIMEOUT as i32;
+
+        match unsafe { twai_transmit(&frame.0, portMAX_DELAY) } {
```

Comment:
If this is supposed to be a non-blocking call, perhaps you should use 0 instead of `portMAX_DELAY`?